### PR TITLE
#484 Drag to select

### PR DESCRIPTION
**Pull request by langonginc**

### CODE REVIEW

File: src/components/svg-canvas-graph.tsx
Comment (thekingofcity) [inline src/components/svg-canvas-graph.tsx:57]: Do we want to support the case where users click select and then start from any nodes? This `handlePointerDown` only works on nodes IMO.

File: src/components/svg-canvas-graph.tsx
Comment (thekingofcity) [inline src/components/svg-canvas-graph.tsx:338]: This could be moved into `svg-wrapper.tsx` if we do not support select from nodes. Also, no more `selectStart` and `selectMoving` need to be shared between this and `svg-wrapper.tsx`.

File: src/util/select-tools.ts
Comment (thekingofcity) [inline src/util/select-tools.ts:8]: `src/util/graph.ts` might be a better place :)

File: src/components/svg-canvas-graph.tsx
Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:338]: But when the mouse go above a node, it will be recorded by handleMoving of canvas rather than it of wrappers.

File: src/util/select-tools.ts
Comment (langonginc) [reply src/util/select-tools.ts:8]: Shall I move all these to graph.tsx?

File: src/components/svg-canvas-graph.tsx
Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:57]: It will be banned.

File: src/components/svg-canvas-graph.tsx
Comment (langonginc) [reply src/components/svg-canvas-graph.tsx]: I have moved it to `svg-wrapper.tsx` .

File: src/util/select-tools.ts
Comment (thekingofcity) [reply src/util/select-tools.ts:8]: Yeah, graph.ts should be an all in one place for all graph-related helper functions :)

Comment (thekingofcity) [reply src/util/select-tools.ts:8]: Also, the function name should be `handleSelectTool` :)

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:202]: there is a setSelected action where you could set selected in batch instead of one by one :)

File: src/util/select-tools.ts
Comment (langonginc) [reply src/util/select-tools.ts]: Done.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:204]: Do we need to clear `selectStart` and `selectMoving`? `<rect />` should not be rendered as long as the mode is not `select`. Or is this used elsewhere?

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: Use `.forEach` can make sure that the nodes are **add** to `selected` continuously rather than clear `selected` and put nodes into it.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:333]: IMO, We may merge those into one `<rect stroke="xxx" strokeWidth="2" fill="xxx" opacity="xxx" />`.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:332]: `id` may not be necessary. Is it being used elsewhere?

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:305]: According to React's best practice, these calculations should be wrapped in `React.useEffect(() => {...}, [...selectStart, ...selectMoving])` as these numbers/functions should only be recalculated when `selectStart` and/or `selectMoving` are changed :)

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:204]: It was inited when press the board in `handleBackgroundDown`

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:201]: Doesn't `handleSelectTool` return all the nodes we want to select? Or does we support first clicking on some nodes with shift and then add more nodes to `selected` by dragging?

Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: HandleSelectTool will return the nodes that we select this time. 

I prefer to select some *group* of nodes without pressing the shift key.

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:332]: No use. I just copied it from somewhere.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:204]: Yeah, I think both of them will be initialized in `handleBackgroundDown` so this line could be removed as long as we do not use them and until we reset them.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:201]: Then if the only scenario we support is to drag to select nodes in the rectangle, why can't we add them in batch via setSelected?

Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: I want to keep the nodes that I have selected last time and then add nodes I select this time.

Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:201]: Okay so I can drag and select nodes 1,2,3... and then drag and select nodes a,b,c... with 1,2,3... still being selected. This could be a feature but I'm wondering if other tools behave like this too. At least Illustrator won't keep previously selected elements.

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:204]: That's right, we can delete them.

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: I know that AutoCAD and Solidworks work like this.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [inline src/components/svg-wrapper.tsx:167]: This could also be removed right? And then `selectStart` and `selectMoving` won't be shared across files :)

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:305]: Should I write `const [SX, setSX] = React.useState(0);` and use `setSX()` in  useEffect()?

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:167]: But, how can our selection-box follow users' mouse?

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx]: Oh no, `selectStart` must be cleared, because the selection box will appear when I click "Select" button before actually select.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:167]: Oh sorry, I commented in the wrong place. It should be `handlePointerMove` in `src\components\svg-canvas-graph.tsx`.

File: src/components/svg-canvas-graph.tsx
Comment (thekingofcity) [inline src/components/svg-canvas-graph.tsx:101]: This could be removed without affecting the feature :)

Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:101]: No, we won't get the pos of mouse when the mouse is on the nodes.

Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:101]: https://github.com/railmapgen/rmp/assets/59787082/694d767e-4974-4558-88d1-fdfaa3ffc093

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: I'll clear the `selected` if the shift key is not pressed.

Comment (langonginc) [reply src/components/svg-wrapper.tsx:201]: I will use addSelected rather than setSelected, because I need make sure the node won't appear twice in `selected`.

File: src/components/svg-canvas-graph.tsx
Comment (thekingofcity) [reply src/components/svg-canvas-graph.tsx:101]: Yes, you're right. Mouse-moving events will be blocked by `e.stopPropagation();` on line 74. I checked the code and looks like if we pop the event up to `handleBackgroundMove` in `svg-wrapper.tsx`, `setSelectMoving` is not required anymore and nothing will break. Correct me if I'm wrong :)

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:308]: Yes, this is more like some _React_ code. Or `const [selectCoord, setSelectCoord] = React.useState({sx:0, sy:0, ex:0, ey:0});` will save you some lines of code :)

File: src/components/svg-canvas-graph.tsx
Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:101]: Yes, that's right. Now this line can be deleted.

File: src/components/svg-wrapper.tsx
Comment (langonginc) [reply src/components/svg-wrapper.tsx:308]: Hmm, I created 4 consts.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:197]: We may clear it when choosing the select tool, but then `selectStart` still needs to be shared between components via runtime-slice. There might be a better way. Let me find it out later :)

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:194]: Using `addSelected` will be O(nm) where n is the size of the current selected and m is the size of nodes returned from `findNodesInRectangle`. The following code should reduce complexity to O(n+m) by creating a set.

```tsx
            const nodesInRectangle = findNodesInRectangle(
                graph.current,
                selectStart.x,
                selectStart.y,
                (x * svgViewBoxZoom) / 100 + svgViewBoxMin.x,
                (y * svgViewBoxZoom) / 100 + svgViewBoxMin.y
            );
            const allSelected = new Set([...selected, ...nodesInRectangle]);
            dispatch(setSelected([...allSelected]));
```

Comment (langonginc) [reply src/components/svg-wrapper.tsx:194]: Oh, I'm not good at object-oriented operations. So, is it $\Theta(n)$ to create a set from an array by `= new Set(simpleArray)`?

Comment (langonginc) [reply src/components/svg-wrapper.tsx:194]: And I think the time complexity of an insert operation in a set is $O(\log n)$.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:197]: You're right, they can not be removed in mouse-up events. Let's keep it as it is.

Comment (langonginc) [reply src/components/svg-wrapper.tsx:197]: I found that `selectMoving` may be cleared as well, or the rectangle will appear once the mouse down. I think there must be something wrong, but I have limited time now.

File: src/components/svg-canvas-graph.tsx
Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:146]: Can we use `React.useEffect()` to clear them when `mode` is changed to `free` from `select`?

Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:146]: So that `selectStart, selectMoving` can be removed from `svg-canvas-graph.tsx`

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx]: No problems. I've got them covered.

File: src/components/svg-wrapper.tsx
Comment (thekingofcity) [reply src/components/svg-wrapper.tsx:194]: Adding one element to a set should be O(1) in JavaScript, at least in v8, as hast table is used as the internal data structure. https://stackoverflow.com/questions/33611509/es6-map-and-set-complexity-v8-implementation

Anyway, even if it is O(logn), the suggested code should be faster than filtering out duplicates using `array.prototype.includes`. That must be O(n) and is only provided as a handy function for a single add, as redux does not allow set, non-serializable objects to be put in.

File: src/components/svg-canvas-graph.tsx
Comment (thekingofcity) [inline src/components/svg-canvas-graph.tsx:143]: If we propagate the pointer up event to svg-wrapper, then this `setMode('free')` could be removed as it will be set in svg-wrapper.

Comment (langonginc) [reply src/components/svg-canvas-graph.tsx:143]: That's right. I will comment e.stopPropagation in this function and remove this 'if'.

File: src/components/panels/tools/tools.tsx
Comment (thekingofcity) [inline src/components/panels/tools/tools.tsx:111]: Do we need to set mode to free if the current mode is select? select mode should get cleared when users click on no matter background/canvas or any node/edge. Another thing is do we want to preserve select mode after a drag-to-selct operation? If so, we may set the mode back to free so that users can return to free mode and drag the background/canvas.

File: src/components/panels/tools/tools.tsx
Comment (thekingofcity) [inline src/components/panels/tools/tools.tsx:119]: Translations need to be added :)

PS: 多选 or 框选 or something elese?

File: src/components/panels/tools/tools.tsx
Comment (langonginc) [reply src/components/panels/tools/tools.tsx:111]: I think it should be cleared when background/canvas/node/edge is clicked.

Comment (langonginc) [reply src/components/panels/tools/tools.tsx:111]: And cleared to free when a selection finishes.

Comment (thekingofcity) [reply src/components/panels/tools/tools.tsx:111]: So we can remove this line? As we can assure users can drag background/canvas/node/edge in select mode by clicking on background/canvas/node/edge.

Comment (langonginc) [reply src/components/panels/tools/tools.tsx:111]: Let me have a try.

~~let me see see~~

Comment (langonginc) [reply src/components/panels/tools/tools.tsx:111]: Hmm, I think it would be better to allow users to click Select button to cancel select operation.

Comment (thekingofcity) [reply src/components/panels/tools/tools.tsx:111]: Okay. We may add some indicators showing the current mode. Maybe darken the select button? Also, make this a one liner like setMode(mode === xxx ? xxx, xxx).

File: src/components/panels/tools/tools.tsx
Comment (langonginc) [reply src/components/panels/tools/tools.tsx:119]: Maybe 多选？

File: src/components/panels/tools/tools.tsx
Comment (langonginc) [reply src/components/panels/tools/tools.tsx:111]: In L116, it is different when the mode is set to "select".

**Review by thekingofcity — APPROVED** [review]
LGTM as long as the commented code is removed :)

**Comment by thekingofcity** [comment]
One more thing, we should add `select` to `RuntimeMode`. Not sure why the type is upcasted to `string`. Maybe `enum` is needed :(

https://github.com/railmapgen/rmp/blob/2a04693a198133e28d055eef6843e9df611bd4a7/src/constants/constants.ts#L80

**Pull request merged**
